companion/src/instruction/bristol.rs

```diff
@@ -6,27 +6,135 @@ use reverie::algebra::RingElement;
 use reverie::Instruction;
 
 use super::Parser;
+use std::str::FromStr;
+
+pub struct BristolHeader {
+    pub n_gate: usize,
+    pub n_wire: usize,
+    pub n_input: usize,
+    pub n_output: usize,
+    pub pending_input: usize,
+    pub pending_output: usize,
+}
+
+impl BristolHeader {
+    fn add_input(&mut self) {
+        self.pending_input -= 1;
+        self.n_input += 1;
+    }
+
+    fn add_output(&mut self) {
+        self.pending_output -= 1;
+        self.n_output += 1;
+    }
+}
+
+impl FromStr for BristolHeader {
+    type Err = std::io::Error;
```

Comment:
Using `std::io::Error` as the error type here is probably fine for now, but this is how you'd do it if you wanted to get rid of all of the incompatible-type `unwrap` calls within:

```rust
pub enum BristolHeaderError {
  Io(std::io::Error),
  ParseError(String),
}
```

Then, you could `impl std::error::Error` and `impl From<std::io::Error> for BristolHeaderError`, and use `BristolHeaderError` as the `Err` type for `impl FromStr`.

(This [thiserror](https://docs.rs/thiserror/1.0.23/thiserror/) crate automates a decent chunk of that).

Reply:
I've partially fixed this, but as `option::NoneError` [doesn't implement](https://github.com/rust-lang/rust/issues/46871) `Error` (for reasons I don't really understand) I can't completely eliminate the `unwrap` calls from `from_str`.

Reply:
Yeah, that's a topic of some contention in Rust.

The "right" way to handle `Option`s in a `Result` context is to `map` the `Some(_)` variant and convert the `None` variant into your own error type. For example:

```rust
let n_gate: usize = parts.next().unwrap().trim().parse()?;
```

becomes (roughly, didn't verify):

```rust
let n_gate: usize = parts.next().map(|p| p.trim()).ok_or_else(|| CountParseError("couldn't parse gate count")).parse()?;
```

where `CountParseError` is a new variant in your `BristolHeaderError` like so:

```rust
CountParseError(&'static str)
```